relay-server/src/actors/project_buffer.rs

```diff
@@ -189,11 +213,130 @@ impl Service for BufferService {
     }
 }
 
-impl Drop for BufferService {
+impl Drop for MemoryBufferService {
     fn drop(&mut self) {
         let count: usize = self.buffer.values().map(|v| v.len()).sum();
         if count > 0 {
             relay_log::error!("dropped queue with {} envelopes", count);
         }
     }
 }
+
+/// [`Buffer`] interface implementation backed by SQLite.
+#[derive(Debug)]
+pub struct SqliteBufferService {
+    db: Pool<Sqlite>,
+}
+
+impl SqliteBufferService {
+    /// Creates a new [`SqliteBufferService`] from the provide path to the sqlite database file.
+    pub async fn from_path(path: PathBuf) -> Result<Self, BufferError> {
+        let options = SqliteConnectOptions::new()
+            .filename(PathBuf::from("sqlite://").join(path))
+            .journal_mode(SqliteJournalMode::Wal)
+            .create_if_missing(true);
+
+        let db = SqlitePoolOptions::new().connect_with(options).await?;
+        Ok(Self { db })
+    }
+
+    /// Handles the enqueueing messages into the internal buffer.
+    async fn handle_enqueue(&self, message: Enqueue) -> Result<(), BufferError> {
+        let Enqueue {
+            key,
+            value: mut managed_envelope,
+        } = message;
+
+        managed_envelope.spool();
+        let envelope_bytes = managed_envelope.take_envelope().to_vec()?;
+        sqlx::query("INSERT INTO envelopes (own_key, sampling_key, envelope) VALUES (?, ?, ?)")
+            .bind(key.own_key.to_string())
+            .bind(key.sampling_key.to_string())
+            .bind(envelope_bytes)
+            .execute(&self.db)
+            .await?;
+        Ok(())
+    }
+
+    /// Handles the dequeueing messages from the internal buffer.
+    ///
+    /// This method removes the envelopes from the buffer and stream them to the sender.
+    async fn handle_dequeue(&self, message: DequeueMany) -> Result<(), BufferError> {
+        let DequeueMany { keys, sender } = message;
+
+        for key in keys {
+            let mut envelopes = sqlx::query(
+                "DELETE FROM envelopes WHERE own_key = ? AND sampling_key = ? RETURNING envelope",
```

Comment:
I think we will need a db index on `(own_key, sampling_key)`.

Reply:
yes, we have an index , see `migrations`